contracts/strategies/DonationVotingOffchain.sol

```diff
@@ -0,0 +1,342 @@
+        internal
+        virtual
+        override
+        onlyActiveAllocation
+    {
+        (address[] memory tokens) = abi.decode(_data, (address[]));
+        uint256 totalNativeAmount;
+
+        for (uint256 i = 0; i < _recipients.length; i++) {
+            if (!_isAcceptedRecipient(_recipients[i])) {
+                revert RECIPIENT_ERROR(_recipients[i]);
+            }
+
+            if (!allowedTokens[tokens[i]] && !allowedTokens[address(0)]) {
+                revert INVALID();
+            }
+
+            // Update the total payout amount for the claim and the total claimable amount
+            amountAllocated[_recipients[i]][tokens[i]] += _amounts[i];
+
+            if (tokens[i] == NATIVE) {
+                totalNativeAmount += _amounts[i];
+            } else {
+                SafeTransferLib.safeTransferFrom(tokens[i], _sender, address(this), _amounts[i]);
+            }
+
+            emit Allocated(_recipients[i], _sender, _amounts[i], abi.encode(tokens[i]));
+        }
+
+        if (msg.value != totalNativeAmount) {
+            revert AMOUNT_MISMATCH();
+        }
+    }
+
+    /// @notice Distributes funds (tokens) to recipients.
+    /// @param _recipientIds The IDs of the recipients
+    /// @param _sender The address of the sender
+    function _distribute(address[] memory _recipientIds, bytes memory, address _sender)
+        internal
+        virtual
+        override
+        onlyAfterAllocation
+    {
+        for (uint256 i; i < _recipientIds.length; i++) {
+            address recipientId = _recipientIds[i];
+
+            uint256 amount = payoutSummaries[recipientId].amount;
+            delete payoutSummaries[recipientId].amount;
+
+            if (amount == 0) {
+                revert INVALID();
+            }
+            poolAmount -= amount;
+
+            address recipientAddress = _recipients[recipientId].recipientAddress;
+            IAllo.Pool memory pool = allo.getPool(poolId);
+            _transferAmount(pool.token, recipientAddress, amount);
+
+            emit Distributed(recipientId, abi.encode(recipientAddress, amount, _sender));
+        }
+    }
+
+    /// @notice Hook called before withdrawing tokens from the pool.
+    function _beforeWithdraw(address, uint256, address) internal virtual override {
+        if (block.timestamp <= allocationEndTime + withdrawalCooldown) revert INVALID();
+    }
+
+    /// @notice Hook called before increasing the pool amount.
+    function _beforeIncreasePoolAmount(uint256) internal virtual override {
+        if (block.timestamp > allocationEndTime) revert POOL_INACTIVE();
+    }
+
+    /// @notice Checks if the timestamps are valid.
+    /// @param _registrationStartTime The start time for the registration
+    /// @param _registrationEndTime The end time for the registration
+    function _isPoolTimestampValid(uint64 _registrationStartTime, uint64 _registrationEndTime)
```

Comment:
Is this being used?

Reply:
It's called from RecipientsExtension::_updatePoolTimestamps https://github.com/allo-protocol/allo-v2/blob/6aa3261364b86c4540d199419af9b3550fa72e8b/contracts/extensions/contracts/RecipientsExtension.sol#L142-L144